### PR TITLE
Schedule guvectorize'd functions over dask.distributed

### PR DESCRIPTION
**Pull request by crusaderky**
Fix failure to pickle a dask array that embeds a guvectorized ufunc in its graph.

- Closes https://github.com/numba/numba/issues/4314
- Closes https://github.com/dask/distributed/issues/3450
- Closes https://github.com/dask/distributed/issues/7929

### CODE REVIEW

File: numba/tests/npyufunc/test_ufuncbuilding.py [hunk 1]
```diff
@@ -317,6 +320,58 @@ def test_guvectorize_invalid_layout(self):
...
+        return self.array
+
+    def tolist(self):
+        return self.array.tolist()
+
+    def __array_ufunc__(self, ufunc, method, *args, **kwargs):
+        if method != "__call__":
+            return NotImplemented
+
+        return NEP13Array(ufunc(*[np.asarray(x) for x in args], **kwargs))
```
Comment (guilhermeleobas) [inline numba/tests/npyufunc/test_ufuncbuilding.py:338]: Is this change needed for the PR to work?

**Review by guilhermeleobas — COMMENTED** [review]
@crusaderky, thanks for your contribution. Do you have a test case using Dask where current code would fail?

File: numba/tests/npyufunc/test_ufuncbuilding.py (hunk 1 above)
Comment (crusaderky) [reply numba/tests/npyufunc/test_ufuncbuilding.py:338]: No, but I thought it would be easier to read if two very similar stub classes were declared next to each other.

**Comment by crusaderky** [comment]
> @crusaderky, thanks for your contribution. Do you have a test case using Dask where current code would fail?

```python
import dask.array as da
import distributed
import numba

client = distributed.Client(processes=True, n_workers=1)

@numba.guvectorize(["f8,f8[:]"], "()->()")
def double(x, out):
    out[:] = x * 2

double(da.arange(5)).compute()
```

File: numba/tests/npyufunc/test_ufuncbuilding.py (hunk 1 above)
Comment (crusaderky) [reply numba/tests/npyufunc/test_ufuncbuilding.py:338]: No;  I felt that it would make it easier to understand the test ccode if these very two similar, and yet subtly different stub classes were presented right next to each other. I can revert if you prefer.

**Review by guilhermeleobas — DISMISSED** [review]
Can confirm the changes fixes the issue.

**Comment by guilhermeleobas** [comment]
@crusaderky, can you add a release notes file? https://numba.readthedocs.io/en/latest/developer/contributing.html#release-notes

**Comment by crusaderky** [comment]
> @crusaderky, can you add a release notes file? https://numba.readthedocs.io/en/latest/developer/contributing.html#release-notes

Done

**Review by guilhermeleobas — APPROVED** [review]
Thanks, @crusaderky

Sorry for the delay on this last review.
